File: Python/01_two_sum.py
# 1. Two Sum

# Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

# You may assume that each input would have exactly one solution, and you may not use the same element twice.

# You can return the answer in any order.

 

# Example 1:

# Input: nums = [2,7,11,15], target = 9
# Output: [0,1]
# Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
# Example 2:

# Input: nums = [3,2,4], target = 6
# Output: [1,2]
# Example 3:

# Input: nums = [3,3], target = 6
# Output: [0,1]
 

# Constraints:

# 2 <= nums.length <= 104
# -109 <= nums[i] <= 109
# -109 <= target <= 109
# Only one valid answer exists.
 

# Follow-up: Can you come up with an algorithm that is less than O(n2) time complexity?


def two_sum(nums, target):
    """
    Find two numbers in the list that add up to the target.

    :param nums: List of integers
    :param target: Integer target sum
    :return: Tuple of indices of the two numbers that add up to the target
    """
    nums_hash = {}
    for i,num in enumerate(nums):
        complement = target-num
        if complement in nums_hash:
            return [nums_hash[complement],i]
        nums_hash[num] = i
    return []



if __name__ == "__main__":
    nums = [2, 7, 11, 15]
    target = 9
    result = two_sum(nums, target)
    print(f"Indices: {result}")  # Output: Indices:]

File: CPP/01_two_sum.cpp
//Two Sum problem in cpp empty function so i can fill it out and main function to test it
#include <iostream>
#include <unordered_map>
#include <utility>  
#include <vector>
using namespace std;

pair<int, int> two_sum(vector<int>& nums, int target) {
    unordered_map<int,int> map;
    for(int i = 0;i<nums.size();i++)
    {
        int complement = target - nums[i];
        if(map.find(complement) != map.end())
        {
            pair<int,int> res = {map[complement], i};
            return res;
        }
        else
        {
            map[nums[i]] = i;
        }        
    }
    return {-1,-1};
}

int main() {
    std::vector<int> numbers = {2, 7, 11, 15};
    int target = 18;
    auto result = two_sum(numbers, target);
    if (result.first != -1) {
        std::cout << "Two sum found at indices: " << result.first << " and " << result.second << std::endl;
    } else {
        std::cout << "No two sum solution found." << std::endl;
    }
    return 0;
}